# Hackaton - iNeuron

## iNeuron

---

A [iNeuron](https://ineuron.ai/home/) é uma empresa de Inteligência Artificial aplicada a soluções comerciais e com um enfático foco em educação com o objetivo de criar melhores soluções de analytics e business para a comunidade. 

Como mencionado acima, também possuem um grande foco na parte da educação, promovendo **hackatons** e novos métodos para melhorar a abordagem educacional.

## O Hackaton

---

Com o objetivo de envolver o público com o aprendizado de máquina, o [Hackaton](https://challenge-ineuron.in/mlchallenge.php#) trás 9 competições diferentes dividas em 3 desafios de aprendizado de máquina. Eles são:

1. ML CHALLENGE 1

    - Back Order Prediction
    - Bankruptcy
    - Restaurant Ratings


2. ML CHALLENGE 2

    - Cardiovascular disease
    - Dota
    - House Rent


3. ML CHALLENGE 3

    - Banking subscription
    - Incident Response Log
    - Kasandr
    
O desafio que vamos explorar é o de **Back Order Prediction**.

## Back Order Prediction

---

Inicialmente, temos que dar o mérito a quem merece.

A competição foi baseada em um [artigo](https://www.researchgate.net/profile/Rodrigo_Santis2/publication/319553365_Predicting_Material_Backorders_in_Inventory_Management_using_Machine_Learning/links/5a383a95a6fdccdd41fdeb99/Predicting-Material-Backorders-in-Inventory-Management-using-Machine-Learning.pdf) de BRASILEIROS (let's go), estes são:

1. Rodrigo Santis - UFMG
2. Eduardo P. de Aguiar - UFJF
3. Leonardo Goliatt - UFJF

## Descrição

---

> **Backorder:** Pedido em atraso.

Um *backorder* é um pedido que, por X motivos, a empresa não conseguiu cumprir.

**Exemplo**, a empresa não conseguiu acompanhar a entrega de um pedido devido a sua alta demanda.

Este *backorder* gera problemas para o cliente, já que seu pedido não chega, logo, a fidelidade com a empresa diminui. E aqui fica a pergunta: 

> **"Qual a empresa quer diminuir a fidelidade com o cliente?"**
>
> Resposta: Se eu fosse qualquer pessoa das finanças eu não queria diminuir essa fidelidade. 



Ainda utilizando o caso de que a empresa está com alta demanda de um produto, dependendo da empresa ela não conseguirá estocar todos estes produtos. Portanto é necessário que a empresa saiba para quais produtos ela pode enfrentar este problema (**esta é a nossa motivação para aplicar o aprendizado de máquina**).

Aqui alguns pontos interessantes que dá para retirar do texto original:

1. "Devido à alta demanda de um produto, a empresa não consegue acompanhar a entrega do pedido". 

2. Encontrar padrões para cada um dos produtos.

## Objetivo

---

Prever se o produto é um *backorder* ou não, isto é, se o produto ficou em atraso ou não.

Variável de cunho binária (0 ou 1).

## Dados 

---

Os dados se referem ao histórico de algumas semanas anterorires à semana que iremos prever. Possui 22 variáveis dependentes (explicativas). Elas são (tenhamos amor a minha vida, não irei traduzir as variáveis e tarefas): 

1. **sku** – Random ID for the product
2. **national_inv** – Current inventory level for the part
3. **lead_time** – Transit time for product (if available)
4. **in_transit_qty** – Amount of product in transit from source
5. **forecast_3_month** – Forecast sales for the next 3 months
6. **forecast_6_month** – Forecast sales for the next 6 months
7. **forecast_9_month** – Forecast sales for the next 9 months
8. **sales_1_month** – Sales quantity for the prior 1 month time period
9. **sales_3_month** – Sales quantity for the prior 3 month time period
10. **sales_6_month** – Sales quantity for the prior 6 month time period
11. **sales_9_month** – Sales quantity for the prior 9 month time period
12. **min_bank** – Minimum recommend amount to stock
13. **potential_issue** – Source issue for part identified
14. **pieces_past_due** – Parts overdue from source
15. **perf_6_month_avg** – Source performance for prior 6 month period
16. **perf_12_month_avg** – Source performance for prior 12 month period
17. **local_bo_qty** – Amount of stock orders overdue
18. **deck_risk** – Part risk flag
19. **oe_constraint** – Part risk flag
20. **ppap_risk** – Part risk flag
21. **stop_auto_buy** – Part risk flag
22. **rev_stop** – Part risk flag
23. **went_on_backorder** – Product actually went on backorder. This is the target value.


## Tasks (Tarefas dadas pelo Hackaton)

---

1.	Do a complete EDA in the python notebook file;
2.	Build a solution design architecture for end to end solution starting from data ingestion to deployment with a detail documentation;
3.	Deploy the end to end automated solution to AWS;
4.	Create a user interface for bulk testing uploaded through excel sheet and for a single record entry both;
5.	Maintain log for each and every prediction request into any database;
6.	Create a low level documentation for end to end solution and deployment;
7.	Define a retraining approach in your documentation;
8.	Create an end to end video of the working of the project.


## Evaluation criteria (Critérios de avaliação)

---

1.	Solution will be evaluated based on the best architecture design of the entire solution;
2.	Solution will be evaluated based on the best EDA;
3.	Solution will be evaluated based on the **best optimization and model accuracy on unseen dataset**;
4.	Solution will be evaluated based on final deployment and testing with a bulk dataset as well as data  with single entry;
5.	Solution will be evaluated based on the best readability and proper commenting of the code;
6.	Solution will be evaluated based on best interactive user interface;
7.	 On the basis of the given data in the features column and target column, you have to build a model which will be able to predict whether an order for a given product can go on backorder or not.


## Timeline (Duração da competição)

---

Vai (ou foi, não sei em que momento você está lendo isso) de 06/05/2020 à 14/06/2020. 

## Prizes (Money, Cash, Dindin, Grana, Premiação)

---

Algo que não iremos ligar muito, pois o objetivo com essa competição é o **aprendizado**.

1. 1st Place - ₹50,000
2. 2nd Place - ₹30,000
3. 3rd Place - ₹20,000

Para quem não sabe, esta moeda (₹) é a Rupia Indiana e, atualmente, custa:

1 Rupia indiana igual a
0,078 Real brasileiro
14 de mai. 14:27 UTC ·

# Considerações iniciais

## Importando as bibliotecas

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Configurações adicionais

In [2]:
plt.style.use("seaborn-muted")
%matplotlib inline
pd.set_option('display.max_columns', None)

## Importando os dados

In [3]:
banco = pd.read_csv("Training_Dataset_v2.csv", low_memory = False)

In [4]:
dados = banco.copy()

In [5]:
dados.head()

sku  national_inv  lead_time  in_transit_qty  forecast_3_month  \
0  1026827           0.0        NaN             0.0               0.0   
1  1043384           2.0        9.0             0.0               0.0   
2  1043696           2.0        NaN             0.0               0.0   
3  1043852           7.0        8.0             0.0               0.0   
4  1044048           8.0        NaN             0.0               0.0   

   forecast_6_month  forecast_9_month  sales_1_month  sales_3_month  \
0               0.0               0.0            0.0            0.0   
1               0.0               0.0            0.0            0.0   
2               0.0               0.0            0.0            0.0   
3               0.0               0.0            0.0            0.0   
4               0.0               0.0            0.0            0.0   

   sales_6_month  sales_9_month  min_bank potential_issue  pieces_past_due  \
0            0.0            0.0       0.0              No              0.0   
1            0.0            0.0       0.0              No              0.0   
2            0.0            0.0       0.0              No              0.0   
3            0.0            0.0       1.0              No              0.0   
4            0.0            4.0       2.0              No              0.0   

   perf_6_month_avg  perf_12_month_avg  local_bo_qty deck_risk oe_constraint  \
0            -99.00             -99.00           0.0        No            No   
1              0.99               0.99           0.0        No            No   
2            -99.00             -99.00           0.0       Yes            No   
3              0.10               0.13           0.0        No            No   
4            -99.00             -99.00           0.0       Yes            No   

  ppap_risk stop_auto_buy rev_stop went_on_backorder  
0        No           Yes       No                No  
1        No           Yes       No                No  
2        No           Yes       No                No  
3        No           Yes       No                No  
4        No           Yes       No                No

In [6]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1687861 entries, 0 to 1687860
Data columns (total 23 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   sku                1687861 non-null  object 
 1   national_inv       1687860 non-null  float64
 2   lead_time          1586967 non-null  float64
 3   in_transit_qty     1687860 non-null  float64
 4   forecast_3_month   1687860 non-null  float64
 5   forecast_6_month   1687860 non-null  float64
 6   forecast_9_month   1687860 non-null  float64
 7   sales_1_month      1687860 non-null  float64
 8   sales_3_month      1687860 non-null  float64
 9   sales_6_month      1687860 non-null  float64
 10  sales_9_month      1687860 non-null  float64
 11  min_bank           1687860 non-null  float64
 12  potential_issue    1687860 non-null  object 
 13  pieces_past_due    1687860 non-null  float64
 14  perf_6_month_avg   1687860 non-null  float64
 15  perf_12_month_avg  1687860 non-n

In [7]:
dados.shape

(1687861, 23)

## Retirar a variável `sku`

---

Como mencionada na descrição desta variável, temos que ela é apenas um ID randômico para cada produto.

In [8]:
dados = dados.drop("sku", axis = 1) 

## Codificar a variável went_on_backorder em 0 e 1

---

In [9]:
columns_yesno = ["potential_issue", "deck_risk", "oe_constraint", "ppap_risk", "stop_auto_buy", "rev_stop", "went_on_backorder"]

dados[columns_yesno] = dados[columns_yesno].replace({"No" : 0, "Yes": 1})

In [10]:
dados["deck_risk"].value_counts()

0.0    1300377
1.0     387483
Name: deck_risk, dtype: int64

## Valores faltantes

---

Como podemos notar, existem exatamente um valor faltante para cada uma das variáveis a seguir:

- Estranho ter somente um valor faltante para grande maioria das variáveis

Vamos identificar onde é esse valor nulo

In [11]:
dados.isnull().sum()

national_inv              1
lead_time            100894
in_transit_qty            1
forecast_3_month          1
forecast_6_month          1
forecast_9_month          1
sales_1_month             1
sales_3_month             1
sales_6_month             1
sales_9_month             1
min_bank                  1
potential_issue           1
pieces_past_due           1
perf_6_month_avg          1
perf_12_month_avg         1
local_bo_qty              1
deck_risk                 1
oe_constraint             1
ppap_risk                 1
stop_auto_buy             1
rev_stop                  1
went_on_backorder         1
dtype: int64

In [12]:
dados = dados.dropna()

# Criar uma baseline

---

A baseline é um modelo simples que tomamos como base pra ver a acurácia dos modelos.

**Se não conseguimos bater um modelo simples, por que fazer um modelo complicado ?**

Um modelo simples a ser utilizado é a Regressão Logística, mas nela vamos ajustar os parâmetros lá dentro, devido ao desbalanceamento da variável target (**went_on_backorder**).

## Separando os dados em treino e teste

---

> **50/50**

ou 

> **70/30**

Melhor optar pelo 70/30 desta vez.

In [13]:
from sklearn.model_selection import train_test_split

x_treino, x_teste, y_treino, y_teste = train_test_split(dados.drop("went_on_backorder", axis = 1), dados["went_on_backorder"], train_size = 0.4, random_state = 1234)

In [14]:
print(x_treino.shape, y_treino.shape, x_teste.shape, y_teste.shape)

(634786, 21) (634786,) (952181, 21) (952181,)


## Aplicando a regressão logística

### Modelo

In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.utils import class_weight

class_weight1 = class_weight.compute_class_weight('balanced', np.unique(y_treino), y_treino)

logreg = LogisticRegression(class_weight = class_weight1, max_iter = 500, random_state = 1234)

logreg.fit(x_treino, y_treino)

C:\Users\Rafael\miniconda3\lib\site-packages\sklearn\linear_model\_logistic.py:856: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  if class_weight == "balanced":
C:\Users\Rafael\miniconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=array([ 0.50347715, 72.39803832]),
                   dual=False, fit_intercept=True, intercept_scaling=1,
                   l1_ratio=None, max_iter=500, multi_class='auto', n_jobs=None,
                   penalty='l2', random_state=1234, solver='lbfgs', tol=0.0001,
                   verbose=0, warm_start=False)

### Previsão para os dados de teste e acurácia

In [16]:
previsao = logreg.predict(x_teste)

logreg.score(x_teste, y_teste) 

0.9930433394491174

### Matriz de confusão

In [17]:
previsao = logreg.predict(x_teste)

from sklearn.metrics import confusion_matrix

confusion_matrix = confusion_matrix(y_teste, previsao)

print(confusion_matrix)

[[945552     44]
 [  6580      5]]


---

Bom, com tudo isso, temos uma baseline com acurácia de 0.9930433394491174. 

Mais pra frente vamos verificar explorar mais as outras métricas:

1. Outras métricas da matriz de confusão;
1. AUC;
2. LogLoss.

**NOW**, we can go to the tasks (não liguem para os textos em inglês aleatórios, é pra treinar).

# Tarefas

---

Dado que temos o que fazer, vamos botar a mão na massa.

## Task 1: Do a complete EDA in the python notebook file

---

Com a informação abaixo, podemos confirmar que o dataset possui 23 variáveis, sendo 22 explicativas e uma 1 independente (target).

Inclusive, não sei vocês perceberam mas a baseline foi feita com base na variável `dados` e todas as transformações também. 

Por isso que lá bem no começo eu salvei uma cópia numa variável chamada `banco` para fazermos todas as modificações novamente e explorar os dados.

![](https://media.giphy.com/media/YnkMcHgNIMW4Yfmjxr/giphy.gif)

### Explorando a variável target (went_on_backorder)

Aqui temos um leve (mentira, grandíssimo) problema de desbalanceamento de classes.

In [18]:
pd.concat([banco["went_on_backorder"].value_counts(), 
                banco["went_on_backorder"].value_counts(normalize = True).mul(100).round(2)], axis = 1, keys = ("Quantidade", "Percentual"))

Quantidade  Percentual
No      1676567       99.33
Yes       11293        0.67

In [19]:
dados.describe().round(2)

national_inv   lead_time  in_transit_qty  forecast_3_month  \
count    1586967.00  1586967.00      1586967.00        1586967.00   
mean         489.51        7.87           45.47            188.74   
std        30461.68        7.06         1309.36           5182.99   
min       -27256.00        0.00            0.00              0.00   
25%            4.00        4.00            0.00              0.00   
50%           14.00        8.00            0.00              0.00   
75%           78.00        9.00            0.00              5.00   
max     12334404.00       52.00       489408.00        1427612.00   

       forecast_6_month  forecast_9_month  sales_1_month  sales_3_month  \
count        1586967.00        1586967.00     1586967.00     1586967.00   
mean             365.34            536.28          56.91         178.48   
std            10099.62          14825.76        1854.77        4971.13   
min                0.00              0.00           0.00           0.00   
25%                0.00              0.00           0.00           0.00   
50%                0.00              0.00           0.00           1.00   
75%               15.00             25.00           5.00          16.00   
max          2461360.00        3777304.00      741774.00     1105478.00   

       sales_6_month  sales_9_month    min_bank  potential_issue  \
count     1586967.00     1586967.00  1586967.00       1586967.00   
mean          352.23         544.13       53.20             0.00   
std          9679.30       15148.71     1119.03             0.02   
min             0.00           0.00        0.00             0.00   
25%             0.00           0.00        0.00             0.00   
50%             3.00           4.00        0.00             0.00   
75%            33.00          50.00        3.00             0.00   
max       2146625.00     3205172.00   205786.00             1.00   

       pieces_past_due  perf_6_month_avg  perf_12_month_avg  local_bo_qty  \
count       1586967.00        1586967.00         1586967.00    1586967.00   
mean              2.17             -1.01              -0.55          0.63   
std             243.40             13.27              11.45         33.44   
min               0.00            -99.00             -99.00          0.00   
25%               0.00              0.69               0.69          0.00   
50%               0.00              0.84               0.82          0.00   
75%               0.00              0.97               0.96          0.00   
max          146496.00              1.00               1.00      12530.00   

       deck_risk  oe_constraint   ppap_risk  stop_auto_buy    rev_stop  \
count  1586967.0     1586967.00  1586967.00     1586967.00  1586967.00   
mean         0.2           0.00        0.12           0.98        0.00   
std          0.4           0.01        0.32           0.16        0.02   
min          0.0           0.00        0.00           0.00        0.00   
25%          0.0           0.00        0.00           1.00        0.00   
50%          0.0           0.00        0.00           1.00        0.00   
75%          0.0           0.00        0.00           1.00        0.00   
max          1.0           1.00        1.00           1.00        1.00   

       went_on_backorder  
count         1586967.00  
mean                0.01  
std                 0.08  
min                 0.00  
25%                 0.00  
50%                 0.00  
75%                 0.00  
max                 1.00